File: aalto/week2/monday/storm.py

def read():
    col = int(input())
    heights = input()
    heights = heights.split(" ")
    return col, heights

def main():
    col, heights = read()
    water = 0
    temp = []
    last_big = 0
    for i in range(col):
        cur = int(heights[i])
        temp.append([cur, last_big])
        if cur > last_big: 
            last_big = cur
        
    last_big = 0
    for i in range(col-1, -1, -1):
        cur = int(heights[i])
        temp2 = temp[i]
        if temp2[1] > last_big:
            temp2[1] = last_big
        water += max(temp2[0], temp2[1])
        if cur > last_big:
            last_big = cur
    print(water)

main()

File: aalto/week2/monday/storm.cpp
#include <iostream>
using namespace std;

int main(){
    int n;
    int water = 0;
    int last_big = 0;
    cin >> n;
    int walls [n];
    int possible [n];
    for (int i = 0; i < n; i++)
    {
        cin >> walls[i];
        possible[i] = last_big;
        if (walls[i] > last_big){
            last_big = walls[i];
        }
    }
    last_big = 0;
    for (int i = n-1; i >= 0; i--)
    {
        if (possible[i] > last_big){
            possible[i] = last_big;
        }
        water += max(walls[i], possible[i]);
        if (walls[i] > last_big){
            last_big = walls[i];
        }
    }
    cout << water << endl;
    return 0;
}